qiskit/circuit/classical/expr/constructors.py

```diff
@@ -300,9 +312,14 @@ def bit_xor(left: typing.Any, right: typing.Any, /) -> Expr:
 
 def _binary_logical(op: Binary.Op, left: typing.Any, right: typing.Any) -> Expr:
     bool_ = types.Bool()
-    left = _coerce_lossless(lift(left), bool_)
-    right = _coerce_lossless(lift(right), bool_)
-    return Binary(op, left, right, bool_)
+    left = lift(left)
+    right = lift(right)
+    try:
+        left = _coerce_lossless(left, bool_)
+        right = _coerce_lossless(right, bool_)
+        return Binary(op, left, right, bool_)
+    except TypeError as ex:
+        raise TypeError(f"invalid types for '{op}': '{left.type}' and '{right.type}'") from ex
```

Comment:
Same comments here about exception handling as before.

That said, `_coerce_lossless` is purely an internal detail, so if you want to modify how  it reports errors (such as `return None` on error and require checking, etc), that's fine by me too.

Reply:
I like that better, I'll go ahead and change it to return `None`.